signimage/signimage.go

```diff
@@ -0,0 +1,357 @@
+			fmt.Errorf("could not get image: http.StatusNotFound")
+		}
+
+		fmt.Errorf("could not get image: %v", err)
+		panic(err)
+	}
+
+	img, err := descriptor.Image()
+	if err != nil {
+		fmt.Errorf("could not Image(): %v", err)
+		panic(err)
+	}
+	fmt.Printf("\n== Successfully pulled image: %s\n", unsignedimagename)
+	fmt.Printf("\n== Looking for files: %s\n", strings.Replace(fileslist, ":", " ", -1))
+
+	/*
+	** loop through all the layers in the image from the top down
+	 */
+	layers, _ := img.Layers()
+	for i := len(layers) - 1; i >= 0; i-- {
+		fmt.Printf("== Searching layer %d\n", i)
+		currentlayer := layers[i]
+		layerreader, err := currentlayer.Uncompressed()
+		if err != nil {
+			fmt.Errorf("could not get layer: %v", err)
+			panic(err)
+		}
+
+		/*
+		** loop through all the files in the layer
+		 */
+		tarreader := tar.NewReader(layerreader)
+		for {
+			header, _ := tarreader.Next()
+			if err == io.EOF || header == nil {
+				break // End of archive
+			}
+
+			// paths in a layer are relative, and supplied paths are absolute so canonicalise
+			canonfilename := canonicalisePath(header.Name)
+			//either the kmod has not yet been found, or we didn't define a list to search for
+			if kmodstosign[canonfilename] == "not found" || (fileslist == "" && kmodstosign[canonfilename] == "") {
+
+				fmt.Printf("\n == Found kmod: %s\n", header.Name)
+				//its a file we wanted and haven't already seen
+				//extract to the local filesystem
+				kmodstosign[canonfilename] = extractFile(extractiondir, header, tarreader)
+				fmt.Printf("\n  == Signing: %s\n", header.Name)
+
+				//sign it
+				signFile(kmodstosign[canonfilename], pubkeyfile, privkeyfile)
+				fmt.Printf("\n  == Signed successfully: %s\n", header.Name)
+
+				// add back in to the new layer
+				err := addToTarball(tarwriter, kmodstosign[canonfilename], header)
+				if err != nil {
+					fmt.Errorf("failed to add %s to layer: %v", canonfilename, err)
+				}
+				fmt.Printf("\n  == Added signed file to new layer: %s\n", header.Name)
+
+			}
+
+		}
+	}
+	missingkmods := 0
+	for k, v := range kmodstosign {
+		if v == "not found" {
+			missingkmods = 1
+			fmt.Printf("Failed to find expected kmod %s \n", k)
+		}
+	}
+	if missingkmods != 0 {
+		fmt.Printf("Failed to find all expected kmods\n")
+		os.Exit(1)
+	}
+
+	//turn our tar archive into a layer
+	mediatype, _ := layers[len(layers)-1].MediaType()
+	signedlayer, err := tarball.LayerFromReader(&b, tarball.WithMediaType(mediatype))
+	if err != nil {
+		fmt.Errorf("failed to generate layer from tar: %v", err)
+	}
+
+	// add the layer to the unsigned image
+	signedimage, err := mutate.AppendLayers(img, signedlayer)
+	if err != nil {
+		fmt.Errorf("failed to append layer: %v", err)
+	}
+
+	fmt.Printf("\n== Appended new layer to image\n")
+
+	if nopush == false {
+		// write the image back to the name:tag set via the envvars
+		signedref, err := name.ParseReference(signedimagename, opts...)
```

Comment:
```suggestion
		signedref, err := name.ParseReference(signedimagename, []name.Option{})
```

I think it is clearer to make sure no one else is using `opts`.